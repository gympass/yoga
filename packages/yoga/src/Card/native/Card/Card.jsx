import React from 'react';
import styled from 'styled-components';
import { string, shape, oneOf, node, bool } from 'prop-types';

import Text from '../../../Text';
import Box from '../../../Box';

const CardShadow = styled(Box).attrs(
  ({
    theme: {
      yoga: {
        components: {
          card: { elevation },
        },
      },
    },
  }) => ({
    elevation,
  }),
)``;

const CardStyled = styled(Box).attrs(
  ({
    variant,
    theme: {
      yoga: {
        colors: { [variant]: color },
        components: { card },
        elevations,
      },
    },
    hasShadow,
  }) => ({
    elevation: hasShadow ? card.elevation : elevations.zero,
    pt: card.padding.top,
    pr: card.padding.right,
    pb: card.padding.bottom,
    pl: card.padding.left,
    bRadius: card.radii,
    bgColor: variant ? color : card.backgroundColor,
    overflow: 'hidden',
  }),
)``;

const Ribbon = styled.View`
  ${({
    variant,
    theme: {
      yoga: {
        components: { card },
        spacing,
        colors: { elements, [variant]: color, white, dark },
      },
    },
  }) => `
    align-self: flex-start;

    margin-bottom: ${card.padding.bottom}px;
    margin-left: -${card.padding.left}px;

    padding:
      ${spacing.xxxsmall}px
      ${spacing.xsmall}px
      ${spacing.xxxsmall}px
      ${spacing.small}px;

    border-top-right-radius: ${card.ribbon.radius}px;
    border-bottom-right-radius: ${card.ribbon.radius}px;
    background-color: ${variant ? color : elements.backgroundAndDisabled};

    color: ${variant ? white : dark};
  `}
`;

const RibbonText = styled(Text.Tiny)`
  ${({
    variant,
    theme: {
      yoga: {
        colors: { white, dark },
      },
    },
  }) => `
    color: ${variant ? white : dark};
  `}
`;

const Card = React.forwardRef(
  (
    { ribbon = {}, children = null, variant = '', hasShadow = true, ...rest },
    ref,
  ) => {
    const CardWrapper = hasShadow ? CardShadow : Box;

    return (
      <CardWrapper>
        <CardStyled ref={ref} {...rest} variant={variant} hasShadow={hasShadow}>
          {Object.keys(ribbon).length > 0 && (
            <Ribbon variant={ribbon.variant}>
              <RibbonText variant={ribbon.variant}>{ribbon.text}</RibbonText>
            </Ribbon>
          )}
          {children}
        </CardStyled>
      </CardWrapper>
    );
  },
);

Card.propTypes = {
  /** text: the content inside the Card Ribbon
   * variant: style the card following the theme (primary, secondary, vibin,
   * hope, energy, relax, peace, verve, uplift, deepPurple, deep, stamina, dark,
   * medium, light, clear, white) */
  ribbon: shape({
    text: string,
    variant: oneOf([
      '',
      'primary',
      'secondary',
      'vibin',
      'hope',
      'energy',
      'relax',
      'peace',
      'verve',
      'uplift',
      'deepPurple',
      'stamina',
      'dark',
      'medium',
      'deep',
      'light',
      'clear',
      'white',
    ]),
  }),
  children: node,
  /** style the card following the theme (primary, secondary, vibin, hope,
   * energy, relax, peace, verve, uplift, deepPurple, deep, stamina, dark,
   * medium, light, clear, white) */
  variant: oneOf([
    '',
    'primary',
    'secondary',
    'vibin',
    'hope',
    'energy',
    'relax',
    'peace',
    'verve',
    'uplift',
    'deepPurple',
    'stamina',
    'dark',
    'medium',
    'deep',
    'light',
    'clear',
    'white',
  ]),
  /** applies a shadow to the card (enabled by default) */
  hasShadow: bool,
};

Card.displayName = 'Card';

export default Card;
