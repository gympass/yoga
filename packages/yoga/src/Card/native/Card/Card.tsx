import React from 'react';
import styled from 'styled-components';
import { string, shape, oneOf, node, InferProps } from 'prop-types';

import Text from '../../../Text/index.native';
import Box from '../../../Box/index.native';
import type { View } from 'react-native';

const CardShadow = styled(Box).attrs(
  ({
    theme: {
      yoga: {
        components: {
          card: { elevation },
        },
      },
    },
  }) => ({
    elevation,
  }),
)``;

const CardStyled = styled(Box).attrs<{ variant: string }>(
  ({
    variant,
    theme: {
      yoga: {
        colors: { [variant]: color },
        components: { card },
      },
    },
  }) => ({
    elevation: card.elevation,
    pt: card.padding.top,
    pr: card.padding.right,
    pb: card.padding.bottom,
    pl: card.padding.left,
    bRadius: card.radii,
    bgColor: variant ? color : card.backgroundColor,
    overflow: 'hidden',
  }),
)``;

const Ribbon = styled.View<{ variant: string; children: React.ReactNode }>`
  ${({
    variant,
    theme: {
      yoga: {
        components: { card },
        spacing,
        colors: { elements, [variant]: color, white, dark },
      },
    },
  }) => `
    align-self: flex-start;

    margin-bottom: ${card.padding.bottom}px;
    margin-left: -${card.padding.left}px;

    padding:
      ${spacing.xxxsmall}px
      ${spacing.xsmall}px
      ${spacing.xxxsmall}px
      ${spacing.small}px;

    border-top-right-radius: ${card.ribbon.radius}px;
    border-bottom-right-radius: ${card.ribbon.radius}px;
    background-color: ${variant ? color : elements.backgroundAndDisabled};

    color: ${variant ? white : dark};
  `}
`;

const RibbonText = styled(Text.Tiny)`
  ${({
    variant,
    theme: {
      yoga: {
        colors: { white, dark },
      },
    },
  }) => `
    color: ${variant ? white : dark};
  `}
`;

const propTypes = {
  /** text: the content inside the Card Ribbon
   * variant: style the card following the theme (primary, secondary, vibin,
   * hope, energy, relax, peace, verve, uplift, deepPurple, deep, stamina, dark,
   * medium, light, clear, white) */
  ribbon: shape({
    text: string,
    variant: oneOf([
      '',
      'primary',
      'secondary',
      'vibin',
      'hope',
      'energy',
      'relax',
      'peace',
      'verve',
      'uplift',
      'deepPurple',
      'stamina',
      'dark',
      'medium',
      'deep',
      'light',
      'clear',
      'white',
    ]),
  }),
  children: node,
  /** style the card following the theme (primary, secondary, vibin, hope,
   * energy, relax, peace, verve, uplift, deepPurple, deep, stamina, dark,
   * medium, light, clear, white) */
  variant: oneOf([
    '',
    'primary',
    'secondary',
    'vibin',
    'hope',
    'energy',
    'relax',
    'peace',
    'verve',
    'uplift',
    'deepPurple',
    'stamina',
    'dark',
    'medium',
    'deep',
    'light',
    'clear',
    'white',
  ]),
};

// TODO: Use system typing
type CardProps = InferProps<typeof propTypes> & { [key: string]: any };

const Card = React.forwardRef<View, CardProps>(
  ({ ribbon, children, ...rest }, ref) => (
    <CardShadow>
      <CardStyled ref={ref} {...rest}>
        {Object.keys(ribbon as {}).length > 0 && (
          <Ribbon variant={ribbon?.variant as string}>
            <RibbonText variant={ribbon?.variant}>{ribbon?.text}</RibbonText>
          </Ribbon>
        )}
        {children}
      </CardStyled>
    </CardShadow>
  ),
);

Card.propTypes = propTypes;

Card.defaultProps = {
  ribbon: {},
  children: null,
  variant: '',
};

Card.displayName = 'Card';

export default Card;
