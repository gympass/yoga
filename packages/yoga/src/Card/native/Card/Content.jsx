import styled from 'styled-components/native';

const Content = styled.View``;

Content.displayName = 'Card.Content';

export default Content;
