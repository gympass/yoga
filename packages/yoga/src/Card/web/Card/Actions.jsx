import styled from 'styled-components';
import { arrayOf, oneOfType } from 'prop-types';

import Button from '../../../Button';
import Tag from '../../../Tag';
import { typeOf } from '../../../shared/propTypes';

const Actions = styled.footer``;

Actions.propTypes = {
  children: oneOfType([
    oneOfType([
      typeOf(Button),
      typeOf(Button.Text),
      typeOf(Button.Outline),
      typeOf(Button.Link),
      typeOf(Tag),
      typeOf(Tag.Informative),
    ]),
    arrayOf(
      oneOfType([
        typeOf(Button),
        typeOf(Button.Text),
        typeOf(Button.Outline),
        typeOf(Button.Link),
        typeOf(Tag),
        typeOf(Tag.Informative),
      ]),
    ),
  ]).isRequired,
};

Actions.displayName = 'Card.Actions';

export default Actions;
