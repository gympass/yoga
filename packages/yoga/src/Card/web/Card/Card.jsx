import { node, oneOf, shape, string, bool } from 'prop-types';
import React from 'react';
import styled from 'styled-components';

import Text from '../../../Text';

const CardStyled = styled.section`
  ${({
    hasShadow,
    variant,
    theme: {
      yoga: {
        spacing,
        colors: { [variant]: color },
        components: { card },
      },
    },
  }) => `
  ${
    hasShadow
      ? `
          padding:
            ${card.padding.top}px
            ${card.padding.right}px
            ${card.padding.bottom}px
            ${card.padding.left}px;
          box-shadow: ${card.elevation};
        `
      : `
          padding: ${spacing.zero}px;
          box-shadow: none;
        `
  }
  border-radius: ${card.radii}px;
  background-color: ${variant ? color : card.backgroundColor};
`}
`;

const Ribbon = styled(Text.Tiny)`
  ${({
    variant,
    theme: {
      yoga: {
        components: { card },
        spacing,
        colors: { elements, [variant]: color, white, text },
      },
    },
  }) => `
  display: inline-block;
  top: ${card.padding.top}px;
  left: 0;

  margin-left: -${card.padding.left}px;
  margin-bottom: ${card.padding.bottom}px;

  padding:
    ${spacing.xxxsmall}px
    ${spacing.xsmall}px
    ${spacing.xxxsmall}px
    ${spacing.small}px;

  border-top-right-radius: ${card.ribbon.radius}px;
  border-bottom-right-radius: ${card.ribbon.radius}px;

  background-color: ${variant ? color : elements.backgroundAndDisabled};

  color: ${variant ? white : text.primary};
  `}
`;

const emptyObj = {};

const Card = React.forwardRef(
  ({ ribbon = emptyObj, children = null, hasShadow = true, ...rest }, ref) => (
    <CardStyled ref={ref} hasShadow={hasShadow} {...rest}>
      {Object.keys(ribbon).length > 0 && (
        <Ribbon variant={ribbon.variant} as="span">
          {ribbon.text}
        </Ribbon>
      )}
      {children}
    </CardStyled>
  ),
);

Card.propTypes = {
  /** text: the content inside the Card Ribbon.
   * variant: style the card following the theme (primary, secondary, vibin, hope,
   * energy, relax, peace, verve, uplift, deepPurple, deep, stamina, dark,
   * medium, light, clear, white) */
  ribbon: shape({
    text: string,
    variant: oneOf([
      '',
      'primary',
      'secondary',
      'vibin',
      'hope',
      'energy',
      'relax',
      'peace',
      'verve',
      'uplift',
      'deepPurple',
      'stamina',
      'dark',
      'medium',
      'deep',
      'light',
      'clear',
      'white',
    ]),
  }),
  children: node,
  /** applies a shadow to the card (enabled by default) */
  hasShadow: bool,
};

Card.displayName = 'Card';

export default Card;
