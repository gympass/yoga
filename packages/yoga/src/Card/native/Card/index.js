import Card from './Card';
import Header from './Header';
import Content from './Content';
import Actions from './Actions';

const ExportCard = Object.assign(Card, {
  Header,
  Content,
  Actions,
});

export default ExportCard;
