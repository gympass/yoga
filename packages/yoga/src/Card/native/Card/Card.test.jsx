import React from 'react';
import { render } from '@testing-library/react-native';
import { Text } from 'react-native';
import { ThemeProvider, Button } from '../../..';
import Card from '.';

describe('<Card />', () => {
  describe('Snapshots', () => {
    it('should match snapshot with default Card', () => {
      const { toJSON } = render(
        <ThemeProvider>
          <Card>
            <Card.Header>
              <Text>Hello World</Text>
            </Card.Header>
            <Card.Content>
              <Text>It&apos;s fine here</Text>
            </Card.Content>
            <Card.Actions>
              <Button>Action&apos;s children must be Button</Button>
            </Card.Actions>
          </Card>
        </ThemeProvider>,
      );

      expect(toJSON()).toMatchSnapshot();
    });
  });
});
