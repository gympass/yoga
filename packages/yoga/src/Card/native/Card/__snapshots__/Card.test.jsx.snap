// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Card /> Snapshots should match snapshot with default Card 1`] = `
<View
  elevation="0 2px 4px      rgba(152, 152, 166, 0.2)"
  style={
    Array [
      Object {
        "shadowColor": "rgba(152, 152, 166, 0.2)",
        "shadowOffset": Object {
          "height": 2,
          "width": 0,
        },
        "shadowOpacity": 1,
        "shadowRadius": 4,
      },
    ]
  }
>
  <View
    bRadius={8}
    bgColor="#FFFFFF"
    elevation="0 2px 4px      rgba(152, 152, 166, 0.2)"
    overflow="hidden"
    pb={16}
    pl={16}
    pr={16}
    pt={16}
    style={
      Array [
        Object {
          "backgroundColor": "#FFFFFF",
          "borderRadius": 8,
          "overflow": "hidden",
          "paddingBottom": 16,
          "paddingLeft": 16,
          "paddingRight": 16,
          "paddingTop": 16,
          "shadowColor": "rgba(152, 152, 166, 0.2)",
          "shadowOffset": Object {
            "height": 2,
            "width": 0,
          },
          "shadowOpacity": 1,
          "shadowRadius": 4,
        },
      ]
    }
    variant=""
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text>
        Hello World
      </Text>
    </View>
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text>
        It's fine here
      </Text>
    </View>
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <View
        accessible={true}
        focusable={false}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "flexDirection": "row",
          }
        }
      >
        <View
          disabled={false}
          full={false}
          inverted={false}
          pressed={false}
          secondary={false}
          small={false}
          style={
            Array [
              Object {
                "alignItems": "center",
                "backgroundColor": "#D8385E",
                "borderRadius": 9999,
                "flexDirection": "row",
                "height": 48,
                "justifyContent": "center",
                "paddingLeft": 24,
                "paddingRight": 24,
              },
            ]
          }
        >
          <Text
            color="#FFFFFF"
            disabled={false}
            inverted={false}
            pressed={false}
            small={false}
            style={
              Array [
                Object {
                  "color": "#FFFFFF",
                  "fontFamily": "Rubik",
                  "fontSize": 16,
                  "fontWeight": "500",
                  "textAlign": "center",
                },
              ]
            }
          >
            Action's children must be Button
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
